prov/gni/src/gnix_msg.c

```diff
@@ -649,6 +687,98 @@ smsg_callback_fn_t gnix_ep_smsg_callbacks[] = {
 	[GNIX_SMSG_T_RNDZV_FIN] = __smsg_rndzv_fin
 };
 
+
+static int __gnix_peek_request(struct gnix_fid_ep *ep,
+		struct gnix_fab_req *req,
+		void *addr,
+		size_t len,
+		void *context,
+		uint64_t flags,
+		uint64_t tag)
+{
+	struct gnix_fid_cq *recv_cq = ep->recv_cq;
+	int rendezvous = !!(req->msg.send_flags & GNIX_MSG_RENDEZVOUS);
+	void *peek_addr = addr;
+	/* all claim work is performed by the tag storage,
+	 * so nothing special here
+	 *
+	 * if no CQ, no data is to be returned. Just inform the user that a
+	 * message is present.
+	 */
+	if (!recv_cq)
+		return FI_SUCCESS;
+
+	/* rendezvous messages on the unexpected queue won't have data
+	 *
+	 * additionally, if the cq format doesn't support passing a buffer
+	 * location and length, then data will not be copied
+	 * */
+
+	if (!rendezvous && peek_addr &&
+			!INVALID_PEEK_FORMAT(recv_cq->attr.format))
+		memcpy(peek_addr, (void *) req->msg.recv_addr,
+				len);
+	else
+		peek_addr = NULL;
+
+	return __recv_completion(ep, req, context, flags, len,
+			peek_addr, req->msg.imm, tag);
+}
+
+static int  __gnix_discard_request(struct gnix_fid_ep *ep,
+		struct gnix_fab_req *req,
+		void *addr,
+		size_t len,
+		void *context,
+		uint64_t flags,
+		uint64_t tag,
+		uint64_t src_addr)
+{
+	int ret;
+	int rendezvous = !!(req->msg.send_flags & GNIX_MSG_RENDEZVOUS);
+
+	GNIX_INFO(FI_LOG_EP_DATA, "discarding req, %p", req);
+
+	if (rendezvous) {
+		/* return a send completion so the sender knows the request/data
+		 * was sent, but discard the data locally
+		 */
+		req->gnix_ep = ep;
+
+		req->msg.recv_addr = (uint64_t) addr;
+		req->msg.recv_len = len;
+		req->user_context = context;
+		req->msg.tag = tag;
+
+		/* TODO: prevent re-lookup of src_addr */
+		ret = _gnix_ep_get_vc(ep, src_addr, &req->vc);
+		if (ret) {
+			GNIX_INFO(FI_LOG_EP_DATA,
+				  "_gnix_ep_get_vc failed: %dn",
+				  ret);
+			return ret;
+		}
+
+		GNIX_INFO(FI_LOG_EP_DATA,
+				"returning rndzv completion for req, %p", req);
+
+
+		/* send completion data. */
+		req->send_fn = __gnix_rndzv_req_send_comp;
+		ret = _gnix_vc_queue_req(req);
+	} else {
+		/* data has already been delivered, so just discard it */
+		ret = __recv_completion(ep, req, context, flags, len,
+				addr, req->msg.imm, tag);
```

Comment:
A discard is always accompanied by either claim or peek. A successful peek should return a CQE, but not necessarily for a claim. I think there should be some open discussion on whether a discard should return a CQE at all, and I would think that it shouldn't. I'll remove it for now.